File: IT/Containers/Esercizi/C++/esercizio18.cpp
/* file esercizio18.cpp */

#include <iostream>
#include <iterator>
#include <algorithm>
#include <vector>

using std::cout;
using std::endl;
using std::string;
using std::ostream_iterator;
using std::vector;
using std::sort;

class Employee {
	private:
		string matr;
		string name;
		string surname;

	public:
		Employee(string, string, string);
		string getMatr() const;
		void setMatr(string matr);
		string getName() const;
		void setName(string name);
		string getSurname() const;
		void setSurname(string surname);

	friend std::ostream& operator<<(std::ostream&, const Employee&);
};

Employee::Employee(string matr, string name, string surname) {
	this->setMatr(matr);
	this->setName(name);
	this->setSurname(surname);
}

string Employee::getMatr() const {
	return this->matr;
}

void Employee::setMatr(string matr) {
	this->matr = matr;
}

string Employee::getName() const {
	return this->name;
}

void Employee::setName(string name) {
	this->name = name;
}

string Employee::getSurname() const {
	return this->surname;
}

void Employee::setSurname(string surname) {
	this->surname = surname;
}

std::ostream& operator<<(std::ostream& os, const Employee& e) {
	os << e.matr << ", " << e.name << " " << e.surname;
	return os;
}

bool comp1(const Employee& e1, const Employee& e2) {
	return e1.getMatr() < e2.getMatr();
}

bool comp2(const Employee& e1, const Employee& e2) {
	return e1.getSurname() < e2.getSurname();
}

bool comp3(const Employee& e1, const Employee& e2) {
	bool c = e1.getName() < e2.getName();
	return c == true ? c : e1.getSurname() < e2.getSurname();
}

int main() {
	vector<Employee> ibm;

	ibm.push_back(Employee("0010", "John", "Von Neumann"));
	ibm.push_back(Employee("0011", "Gene", "Amdahl"));
	ibm.push_back(Employee("0001", "Thomas", "Watson"));
	ibm.push_back(Employee("0101", "Bob", "Bemer"));
	ibm.push_back(Employee("0100", "John", "Backus"));

	cout << "Lista originale:" << endl;

	cout << "[";
	copy(ibm.begin(), ibm.end(), ostream_iterator<Employee>(cout, ", "));
	cout << "\b\b]" << endl;

	cout << "Ordinata per matricola:" << endl;

	sort(ibm.begin(), ibm.end(), comp1);

	cout << "[";
	copy(ibm.begin(), ibm.end(), ostream_iterator<Employee>(cout, ", "));
	cout << "\b\b]" << endl;

	cout << "Ordinata per cognome:" << endl;

	sort(ibm.begin(), ibm.end(), comp2);

	cout << "[";
	copy(ibm.begin(), ibm.end(), ostream_iterator<Employee>(cout, ", "));
	cout << "\b\b]" << endl;

	cout << "Ordinata per nome e per cognome:" << endl;

	sort(ibm.begin(), ibm.end(), comp3);

	cout << "[";
	copy(ibm.begin(), ibm.end(), ostream_iterator<Employee>(cout, ", "));
	cout << "\b\b]" << endl;
}

File: IT/1/Containers/Esercizi/Java/Esercizio18.java
/* file Esercizio18.java */

import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

class Employee implements Comparable<Employee> {
	private String matr;
	private String name;
	private String surname;

	public Employee(String matr, String name, String surname) {
		this.setMatr(matr);
		this.setName(name);
		this.setSurname(surname);
	}

	public String getMatr() {
		return this.matr;
	}

	public void setMatr(String matr) {
		this.matr = matr;
	}

	public String getName() {
		return this.name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getSurname() {
		return this.surname;
	}

	public void setSurname(String surname) {
		this.surname = surname;
	}

	public int compareTo(Employee e) {
		return this.matr.compareTo(e.getMatr());
	}

	@Override
	public String toString() {
		return matr + ", " + name + " " + surname;
	}
}

public class Esercizio18 {
	public static void main(String [] args) {
		List<Employee> ibm = new ArrayList<Employee>();

		ibm.add(new Employee("0010", "John", "Von Neumann"));
		ibm.add(new Employee("0011", "Gene", "Amdahl"));
		ibm.add(new Employee("0001", "Thomas", "Watson"));
		ibm.add(new Employee("0101", "Bob", "Bemer"));
		ibm.add(new Employee("0100", "John", "Backus"));

		System.out.println("Lista originale:");
		System.out.println(ibm);

		System.out.println("Ordinata per matricola:");

		Collections.sort(ibm);
		System.out.println(ibm);

		System.out.println("Ordinata per cognome:");

		Collections.sort(ibm, new Comparator<Employee>() {
			@Override
			public int compare(Employee a, Employee b) {
				return a.getSurname().compareTo(b.getSurname());
			}
		});

		System.out.println(ibm);

		System.out.println("Ordinata per nome e per cognome:");

		Collections.sort(ibm, new Comparator<Employee>() {
			@Override
			public int compare(Employee a, Employee b) {
				int c = a.getName().compareTo(b.getName());
				return c != 0 ? c : a.getSurname().compareTo(b.getSurname());
			}
		});

		System.out.println(ibm);
	}
}

